### ISSUE TITLE
Rolling deli into IDB

### ISSUE THREAD

**Issue by LucaCappelletti94** [devashishdxt/deli#5]
Now that deli is aligned with IDB, should we roll deli into IDB as you mentioned [here](https://github.com/devashishdxt/deli/issues/2#issuecomment-2373317391) before proceeding with extending features like the composite primary keys?

If so, what do you think should be done?

**Comment by devashishdxt** [comment]
I think we should first try to implement all the indexed db features in deli and iterate over the API until we're satisfied. Then, we can merge it with `idb`.

As mentioned, some of the missing features are:

- Composite primary keys
- Indexes with multiple keys

**Comment by LucaCappelletti94** [comment]
Ok!

**Comment by LucaCappelletti94** [comment]
Now the features we wanted to add are merged in. How would you rather we proceed?

**Comment by ifiokjr** [comment]
Also it doesn't look like there's been a recent release of deli? Can it be published to [`crates.io`](https://crates.io/crates/deli)?

**Comment by devashishdxt** [comment]
Yes. I'll release it after a few small updates over the weekend.

**Comment by LucaCappelletti94** [comment]
@devashishdxt regarding the merger with idb?

**Comment by devashishdxt** [comment]
@LucaCappelletti94 Firstly, thanks a lot for your contributions. But, there are still a lot of rough edges in this crate. So, I don't think we can merge it with `idb` right now.

For example:

- Indexed DB supports composite indexes, composite unique indexes and composite multi-entry indexes. There's no way to specify this right now.
- For composite keys, there's no compile-time validation if the user specifies one key as auto-increment (which is not supported by indexed db in case of composite keys).

It is possible to handle all these (and more) edge cases using derive macros. But, until then, I'd like to keep this as a separate crate.

**Comment by LucaCappelletti94** [comment]
I am not sure I understood the cases you describe, it was my understanding that I had added support for these in the latest PR - could you provide pseudocode examples of them?

**Comment by devashishdxt** [comment]
> I am not sure I understood the cases you describe, it was my understanding that I had added support for these in the latest PR - could you provide pseudocode examples of them?

```rust
struct MyStruct {
    pub a: u32,
    pub b: String,
    pub c: Vec<String>,
}
```

If I want two composite indexes on this struct. For example,an index on (a,b) and (b,c) with multi-entry. I can't do that right now.

**Comment by LucaCappelletti94** [comment]
Ok so, not multiple primary keys (which I would not know how to represent in an SQL analogy) but something like multiple unique constraints?

```sql
CREATE TABLE my_table (
    id SERIAL PRIMARY KEY,
    column1 TEXT NOT NULL,
    column2 TEXT NOT NULL,
    column3 TEXT,
    UNIQUE (column1, column2),
    UNIQUE (column2, column3)
);
```

**Comment by devashishdxt** [comment]
> Ok so, not multiple primary keys (which I would not know how to represent in an SQL analogy) but something like multiple unique constraints?
> 
> 
> 
> ```sql
> 
> CREATE TABLE my_table (
>     id SERIAL PRIMARY KEY,
>     column1 TEXT NOT NULL,
>     column2 TEXT NOT NULL,
>     column3 TEXT,
>     UNIQUE (column1, column2),
>     UNIQUE (column2, column3)
> );
> ```

Yes, in SQL terms, you can think of it as multiple unique constraints.

**Comment by LucaCappelletti94** [comment]
Kk, on it.

Btw the following is actually already supported by the PR #6, I just don't know how to add a test suite bit for that as it stops the compilation of course.

> For composite keys, there's no compile-time validation if the user specifies one key as auto-increment (which is not supported by indexed db in case of composite keys).

**Comment by devashishdxt** [comment]
@LucaCappelletti94 I managed to find some time and look into it. This (#8) fixes all the issues.

**Comment by LucaCappelletti94** [comment]
Oh okay, I was also working on it - did you also add the getter methods for the composite indexes? I honestly noticed I had not added them in my previous PR for the composite primary keys.

**Comment by devashishdxt** [comment]
> Oh okay, I was also working on it - did you also add the getter methods for the composite indexes? I honestly noticed I had not added them in my previous PR for the composite primary keys.

I've made major changes in how the derive macros are implemented. I'm not sure what you mean by getter methods for composite indexes but please have a look in the documentation and create an issue if you have any feedback of something is missing.

**Comment by LucaCappelletti94** [comment]
Ok! Is the full test suite running? I noticed inly yesterday that by creating a second test file it was only running that second one, instead of the complete test suite, and the web test was actually failing

**Comment by devashishdxt** [comment]
> Ok! Is the full test suite running? I noticed inly yesterday that by creating a second test file it was only running that second one, instead of the complete test suite, and the web test was actually failing

I've added GitHub actions to run the tests. There are currently not enough tests. I'll probably add some over the next weekend.

**Comment by devashishdxt** [comment]
Also, documentation on how to use `Cursor` and `KeyCursor` is missing.

**Comment by LucaCappelletti94** [comment]
Are you able to add a test coverage with something like coveralls? If so, I could take care of adding more tests myself with that.

**Comment by devashishdxt** [comment]
> Are you able to add a test coverage with something like coveralls? If so, I could take care of adding more tests myself with that.

Last time I checked, there's no code coverage tool that works with `wasm-bindgen-test` (running tests on browsers). I'll check and add if there are any tools now.

**Comment by LucaCappelletti94** [comment]
Have you already added the composite unique indices in pull request #8 ?

**Comment by devashishdxt** [comment]
> Have you already added the composite unique indices in pull request #8 ?

Yes

**Comment by LucaCappelletti94** [comment]
I am not sure I understand what is the expected decorator syntax to define one, could you kindly write an example or point me to it?

**Comment by LucaCappelletti94** [comment]
Ah no my bad, found it. For future reference, it is:

```rust
use deli::Model;
use serde::{Deserialize, Serialize};

#[derive(Serialize, Deserialize, Model)]
#[deli(unique(fields(id, permissions)))] // <- This defines a composite multi-entry index on the `id` and `permissions` fields
pub struct Employee {
    #[deli(auto_increment)]
    id: u32,
    permissions: Vec<String>,
}
```

**Comment by LucaCappelletti94** [comment]
I also see that you have already implemented the getter methods for the composite indices I had mentioned earlier, perfect!

**Comment by devashishdxt** [comment]
Closing this issue for now. Create an issue in `idb`: https://github.com/devashishdxt/idb/issues/32
